File: tests/scanner.test.ts
import { mkdirSync, mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";

import { afterEach, describe, expect, it } from "vitest";

import { scanProject } from "../src/project/scanner.js";

const tempDirs: string[] = [];

afterEach(() => {
  while (tempDirs.length > 0) {
    const dir = tempDirs.pop();

    if (dir) {
      rmSync(dir, { force: true, recursive: true });
    }
  }
});

describe("scanProject", () => {
  it("respects ignore files and default ignore directories", async () => {
    const rootDir = createWorkspace({
      ".gitignore": "coverage/\nignored.ts\n",
      ".deepvibeignore": "private/\n",
      "src/api/request.ts": "export const request = true;\n",
      "src/api/request.test.ts": "export const requestTest = true;\n",
      "ignored.ts": "ignore me\n",
      "coverage/out.txt": "ignore me\n",
      "private/secret.ts": "ignore me\n",
      "node_modules/pkg/index.js": "ignore me\n",
      "dist/bundle.js": "ignore me\n"
    });

    const result = await scanProject({
      rootDir,
      instruction: "update request timeout"
    });

    expect(result.scannedFiles).toBe(4);
    expect(result.candidates).toContain("src/api/request.ts");
    expect(result.candidates).not.toContain("ignored.ts");
    expect(result.candidates).not.toContain("private/secret.ts");
  });

  it("forces explicit @file references into the candidate list", async () => {
    const rootDir = createWorkspace({
      "src/services/api.ts": "export const api = true;\n",
      "src/features/auth/login.ts": "export const login = true;\n",
      "README.md": "# demo\n"
    });

    const result = await scanProject({
      rootDir,
      instruction: "refactor auth flow @src/features/auth/login.ts",
      maxCandidates: 1,
      recentGitFiles: ["src/services/api.ts"]
    });

    expect(result.explicitPaths).toEqual(["src/features/auth/login.ts"]);
    expect(result.candidates[0]).toBe("src/features/auth/login.ts");
    expect(result.candidates).toContain("src/services/api.ts");
  });

  it("uses recent git files and keyword matches to produce stable ordering", async () => {
    const rootDir = createWorkspace({
      "src/api/request.ts": "export const request = true;\n",
      "src/api/request.test.ts": "export const requestTest = true;\n",
      "src/auth/login.ts": "export const login = true;\n",
      "docs/notes.md": "notes\n"
    });

    const result = await scanProject({
      rootDir,
      instruction: "update request flow",
      recentGitFiles: ["src/auth/login.ts"]
    });

    expect(result.candidates.slice(0, 3)).toEqual([
      "src/api/request.test.ts",
      "src/api/request.ts",
      "src/auth/login.ts"
    ]);
  });
});

function createWorkspace(files: Record<string, string>): string {
  const rootDir = mkdtempSync(path.join(tmpdir(), "deepvibe-scanner-"));
  tempDirs.push(rootDir);

  for (const [relativePath, contents] of Object.entries(files)) {
    const filePath = path.join(rootDir, relativePath);

    mkdirSync(path.dirname(filePath), { recursive: true });
    writeFileSync(filePath, contents, "utf8");
  }

  return rootDir;
}

File: internal/scanner/scanner_test.go
package scanner

import (
	"context"
	"os"
	"path/filepath"
	"testing"
)

func TestScanProjectFindsExplicitPathsAndIgnoresFiles(t *testing.T) {
	root := t.TempDir()
	writeTestFile(t, filepath.Join(root, "src", "config.ts"), "export const value = 1\n")
	writeTestFile(t, filepath.Join(root, "src", "engine.ts"), "export const engine = 1\n")
	writeTestFile(t, filepath.Join(root, "tests", "engine.test.ts"), "test('engine')\n")
	writeTestFile(t, filepath.Join(root, "dist", "bundle.js"), "ignored\n")
	writeTestFile(t, filepath.Join(root, ".deepvibeignore"), "ignored/**\n")
	writeTestFile(t, filepath.Join(root, "ignored", "secret.ts"), "ignored\n")

	result, err := ScanProject(context.Background(), Options{
		RootDir:       root,
		Instruction:   "update @src/config.ts and engine behavior",
		MaxCandidates: 2,
	})
	if err != nil {
		t.Fatalf("ScanProject returned error: %v", err)
	}

	if len(result.ExplicitPaths) != 1 || result.ExplicitPaths[0] != "src/config.ts" {
		t.Fatalf("expected explicit src/config.ts, got %#v", result.ExplicitPaths)
	}
	if result.ScannedFiles != 3 {
		t.Fatalf("expected 3 scanned files, got %d", result.ScannedFiles)
	}
	if len(result.Candidates) == 0 || result.Candidates[0] != "src/config.ts" {
		t.Fatalf("expected explicit path first, got %#v", result.Candidates)
	}
}

func writeTestFile(t *testing.T, path string, contents string) {
	t.Helper()
	if err := os.MkdirAll(filepath.Dir(path), 0755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(path, []byte(contents), 0644); err != nil {
		t.Fatal(err)
	}
}
